source/AI.cpp

```diff
@@ -651,12 +651,15 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 			continue;
 		}
 
-		// Run away if your hostile target is not disabled and you are badly damaged.
+		// Run away if your hostile target is not disabled
+		// and you are either badly damaged or have run out of ammo.
 		// Player ships never stop targeting hostiles, while hostile mission NPCs will
 		// do so only if they are allowed to leave.
-		const bool shouldFlee = (personality.IsFleeing() ||
-			(!personality.IsDaring() && !personality.IsStaying()
-			&& healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward()));
+		const bool lowHealth = healthRemaining < RETREAT_HEALTH + .25 * personality.IsCoward();
+		const bool needsAmmo = it->GetAICache().NeedsAmmo();
+		const bool shouldFlee = personality.IsFleeing()
+				|| (!personality.IsStaying()
+				&& ((!personality.IsDaring() && lowHealth) || needsAmmo));
```

Comment:
Can you move this in a separate anonymous namespace function? and fix the conflicts and then this should be good to go :)

Reply:
Do you mean `shouldFlee` or the fleeing procedure?

Reply:
`shouldFlee`

Reply:
Do what StayOrLinger() does for shouldStay.